### PR TITLE
flow: Add openstack discovery component

### LINKED ISSUE #2278: Flow: discovery.openstack component
For feature parity, we should bring over the upstream Prometheus SD mechanism and introduce it as a Flow component.

### PR DESCRIPTION
**Pull request by marctc**
#### PR Description

This PR adds `discovery.openstack` service discovery component.

#### Which issue(s) this PR fixes

Fixes #2278 

#### PR Checklist

- [x] CHANGELOG updated
- [x] Documentation added
- [x] Tests updated

### CODE REVIEW

File: docs/sources/flow/reference/components/discovery.openstack.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+
+[OpenStack]: https://docs.openstack.org/nova/latest/
+
+## Usage
+
+```river
+discovery.openstack "LABEL" {
+  role   = "hypervisor"
+	region = "us-east-1"
+}
```
Comment (spartan0x117) [inline docs/sources/flow/reference/components/discovery.openstack.md:18]:
```suggestion
discovery.openstack "LABEL" {
  role   = "hypervisor"
  region = "us-east-1"
}
```

File: docs/sources/flow/reference/components/discovery.openstack.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+The following arguments are supported:
+
+Name                | Type       | Description                                                            | Default              | Required
+------------------- | ---------- | ---------------------------------------------------------------------- | -------------------- | --------
+`role`              | `string`   | Role of the discovered targets.                                        |                      | yes
+`region`            | `string`   | OpenStack region.                                                      |                      | yes
+`identity_endpoint` | `string`   | Specifies the HTTP endpoint that is required to work with te Identity API of the appropriate version | | no
+`username`          | `string`   | OpenStack username for the Identity V2 and V3 APIs.                    |                      | no
+`userid`            | `string`   | OpenStack userid for the Identity V2 and V3 APIs.                      |                      | no
+`password`          | `secret`   | password for the Identity V2 and V3 APIs.                              |                      | no
```
Comment (spartan0x117) [inline docs/sources/flow/reference/components/discovery.openstack.md:32]: nit:
```suggestion
`password`          | `secret`   | Password for the Identity V2 and V3 APIs.                              |                      | no
```

File: docs/sources/flow/reference/components/discovery.openstack.md [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+`project_id` and `project_name` fields are optional for the Identity V2 API. Some providers allow you to specify a `project_name` instead of the `project_id`. Some require both.
+
+`application_credential_id` or `application_credential_name` fields are required if using an application credential to authenticate. Some providers allow you to create an application credential to authenticate rather than a password.
+
+`application_credential_secret` field is required if using an application credential to authenticate.
+
+`all_tenants` is only relevant for the `instance` role and usually requires admin permissions.
+
+`availability` must be one of `public`, `admin` or `internal`.
```
Comment (spartan0x117) [inline docs/sources/flow/reference/components/discovery.openstack.md:57]:
```suggestion
`availability` must be one of `public`, `admin`, or `internal`.
```

**Review by spartan0x117 — APPROVED** [review]
I added a few comments for the docs, but otherwise LGTM.

File: docs/sources/flow/reference/components/discovery.openstack.md [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+`identity_endpoint` | `string`   | Specifies the HTTP endpoint that is required to work with te Identity API of the appropriate version | | no
+`username`          | `string`   | OpenStack username for the Identity V2 and V3 APIs.                    |                      | no
+`userid`            | `string`   | OpenStack userid for the Identity V2 and V3 APIs.                      |                      | no
+`password`          | `secret`   | password for the Identity V2 and V3 APIs.                              |                      | no
+`domain_name`       | `string`   | OpenStack domain name for the Identity V2 and V3 APIs.                 |                      | no
+`domain_id`         | `string`   | OpenStack domain id for the Identity V2 and V3 APIs.                   |                      | no
+`project_name`      | `string`   | OpenStack project name for the Identity V2 and V3 APIs.                |                      | no
+`project_id`        | `string`   | OpenStack project id for the Identity V2 and V3 APIs.                  |                      | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/discovery.openstack.md:36]:
```suggestion
`domain_id`         | `string`   | OpenStack domain ID for the Identity V2 and V3 APIs.                   |                      | no
`project_name`      | `string`   | OpenStack project name for the Identity V2 and V3 APIs.                |                      | no
`project_id`        | `string`   | OpenStack project ID for the Identity V2 and V3 APIs.                  |                      | no
```

File: docs/sources/flow/reference/components/discovery.openstack.md [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+`application_credential_name` | `string`   | OpenStack application credential name for the Identity V2 and V3 APIs.   |          | no
+`application_credential_id`   | `string`   | OpenStack application credential id for the Identity V2 and V3 APIs.     |          | no
```
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/discovery.openstack.md:38]:
```suggestion
`application_credential_id`   | `string`   | OpenStack application credential ID for the Identity V2 and V3 APIs.     |          | no
```

**Pull request merged**